File: TS/src/stream/handler.ts
// Package stream provides streaming downloads with range support

import { Format } from "../types/format.js";

// Handler manages streaming downloads with range support
export class Handler {
    private UserAgent: string;
    private ChunkSize: number;
    private MaxRetries: number;

    constructor() {
        this.UserAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";
        this.ChunkSize = 10 * 1024 * 1024; // 10MB chunks
        this.MaxRetries = 3;
    }

    // SetUserAgent sets the user agent for requests
    SetUserAgent(userAgent: string): void {
        this.UserAgent = userAgent;
    }

    // SetChunkSize sets the chunk size for range downloads
    SetChunkSize(size: number): void {
        this.ChunkSize = size;
    }

    // SetMaxRetries sets the maximum number of retries
    SetMaxRetries(retries: number): void {
        this.MaxRetries = retries;
    }

    // Download downloads a format and returns the data as a Buffer
    async Download(format: Format, signal?: AbortSignal): Promise<Buffer> {
        if (!format.URL) {
            throw new Error("format has no URL");
        }

        // For formats with known content length, use range requests
        if (format.ContentLength > 0) {
            return this.downloadWithRanges(format.URL, 0, format.ContentLength, signal);
        }

        // Otherwise, simple download
        return this.downloadSimple(format.URL, signal);
    }

    // DownloadRange downloads a specific byte range
    async DownloadRange(format: Format, start: number, end: number, signal?: AbortSignal): Promise<Buffer> {
        if (!format.URL) {
            throw new Error("format has no URL");
        }

        return this.downloadWithRanges(format.URL, start, end, signal);
    }

    // GetStream returns a readable stream for the format
    async GetStream(format: Format, signal?: AbortSignal): Promise<{ body: ReadableStream<Uint8Array>; contentLength: number }> {
        return this.GetStreamRange(format, 0, -1, signal);
    }

    // GetStreamRange returns a readable stream for a specific byte range
    async GetStreamRange(format: Format, start: number, end: number, signal?: AbortSignal): Promise<{ body: ReadableStream<Uint8Array>; contentLength: number }> {
        if (!format.URL) {
            throw new Error("format has no URL");
        }

        const headers: Record<string, string> = this.getHeaders();

        // Set range header
        if (end > 0) {
            headers["Range"] = `bytes=${start}-${end}`;
        } else if (start > 0) {
            headers["Range"] = `bytes=${start}-`;
        }

        const resp = await fetch(format.URL, {
            method: "GET",
            headers,
            signal,
        });

        if (resp.status !== 200 && resp.status !== 206) {
            throw new Error(`unexpected status: ${resp.status}`);
        }

        const contentLength = parseInt(resp.headers.get("Content-Length") || "0", 10);

        if (!resp.body) {
            throw new Error("response has no body");
        }

        return { body: resp.body, contentLength };
    }

    // downloadWithRanges downloads using chunked range requests
    private async downloadWithRanges(url: string, start: number, end: number, signal?: AbortSignal): Promise<Buffer> {
        const chunks: Buffer[] = [];
        let downloaded = start;

        while (downloaded < end) {
            if (signal?.aborted) {
                throw new Error("Download aborted");
            }

            let chunkEnd = downloaded + this.ChunkSize - 1;
            if (chunkEnd >= end) {
                chunkEnd = end - 1;
            }

            const chunk = await this.downloadChunk(url, downloaded, chunkEnd, signal);
            chunks.push(chunk);

            downloaded = chunkEnd + 1;
        }

        return Buffer.concat(chunks);
    }

    // downloadChunk downloads a single chunk with retries
    private async downloadChunk(url: string, start: number, end: number, signal?: AbortSignal): Promise<Buffer> {
        let lastErr: Error | null = null;

        for (let attempt = 0; attempt < this.MaxRetries; attempt++) {
            if (attempt > 0) {
                if (signal?.aborted) {
                    throw new Error("Download aborted");
                }
                await sleep(attempt * 1000);
            }

            try {
                return await this.doChunkRequest(url, start, end, signal);
            } catch (err) {
                lastErr = err instanceof Error ? err : new Error(String(err));
            }
        }

        throw new Error(`failed after ${this.MaxRetries} retries: ${lastErr?.message}`);
    }

    // doChunkRequest performs a single chunk request
    private async doChunkRequest(url: string, start: number, end: number, signal?: AbortSignal): Promise<Buffer> {
        const headers = this.getHeaders();
        headers["Range"] = `bytes=${start}-${end}`;

        const resp = await fetch(url, {
            method: "GET",
            headers,
            signal,
        });

        if (resp.status !== 206 && resp.status !== 200) {
            throw new Error(`unexpected status: ${resp.status}`);
        }

        const arrayBuffer = await resp.arrayBuffer();
        return Buffer.from(arrayBuffer);
    }

    // downloadSimple performs a simple download without range requests
    private async downloadSimple(url: string, signal?: AbortSignal): Promise<Buffer> {
        const headers = this.getHeaders();

        const resp = await fetch(url, {
            method: "GET",
            headers,
            signal,
        });

        if (resp.status !== 200) {
            throw new Error(`unexpected status: ${resp.status}`);
        }

        const arrayBuffer = await resp.arrayBuffer();
        return Buffer.from(arrayBuffer);
    }

    // getHeaders returns required headers for requests
    private getHeaders(): Record<string, string> {
        return {
            "User-Agent": this.UserAgent,
            "Accept": "*/*",
            "Accept-Language": "en-US,en;q=0.9",
            "Origin": "https://www.youtube.com",
            "Referer": "https://www.youtube.com/",
        };
    }

    // GetStreamInfo fetches stream metadata without downloading
    async GetStreamInfo(format: Format): Promise<StreamInfo> {
        if (!format.URL) {
            throw new Error("format has no URL");
        }

        const headers = this.getHeaders();

        const resp = await fetch(format.URL, {
            method: "HEAD",
            headers,
        });

        if (resp.status !== 200) {
            throw new Error(`unexpected status: ${resp.status}`);
        }

        const contentLength = parseInt(resp.headers.get("Content-Length") || "0", 10);
        const contentType = resp.headers.get("Content-Type") || "";
        const acceptRanges = resp.headers.get("Accept-Ranges") === "bytes";

        return {
            ContentLength: contentLength,
            ContentType: contentType,
            AcceptRanges: acceptRanges,
        };
    }

    // DownloadWithProgress downloads with progress reporting
    async DownloadWithProgress(format: Format, callback: ProgressCallback, signal?: AbortSignal): Promise<Buffer> {
        if (!format.URL) {
            throw new Error("format has no URL");
        }

        let total = format.ContentLength;
        if (total <= 0) {
            try {
                const info = await this.GetStreamInfo(format);
                if (info.ContentLength > 0) {
                    total = info.ContentLength;
                }
            } catch {
                // Ignore errors getting stream info
            }
        }

        const chunks: Buffer[] = [];
        let downloaded = 0;
        const startTime = Date.now();

        if (total > 0) {
            while (downloaded < total) {
                if (signal?.aborted) {
                    throw new Error("Download aborted");
                }

                let chunkEnd = downloaded + this.ChunkSize - 1;
                if (chunkEnd >= total) {
                    chunkEnd = total - 1;
                }

                const chunk = await this.downloadChunk(format.URL, downloaded, chunkEnd, signal);
                chunks.push(chunk);

                downloaded += chunk.length;

                const elapsed = (Date.now() - startTime) / 1000;
                const speed = downloaded / elapsed;

                callback({
                    Total: total,
                    Downloaded: downloaded,
                    Speed: speed,
                });
            }

            return Buffer.concat(chunks);
        }

        // Simple download for unknown length
        const data = await this.downloadSimple(format.URL, signal);
        callback({
            Total: data.length,
            Downloaded: data.length,
            Speed: data.length / ((Date.now() - startTime) / 1000),
        });
        return data;
    }
}

// StreamInfo contains metadata about a stream
export interface StreamInfo {
    ContentLength: number;
    ContentType: string;
    AcceptRanges: boolean;
}

// Progress tracks download progress
export interface Progress {
    Total: number;
    Downloaded: number;
    Speed: number; // bytes per second
}

// ProgressCallback is called with download progress updates
export type ProgressCallback = (progress: Progress) => void;

// NewHandler creates a new stream handler with default settings
export function NewHandler(): Handler {
    return new Handler();
}

// Helper function
function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
}

File: V2/stream/handler.go
package stream

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"

	"github.com/elucid503/overture-play/v2/types"
)

// Handler manages streaming downloads with range support
type Handler struct {
	HTTPClient *http.Client
	UserAgent  string

	ChunkSize  int64
	MaxRetries int
}

// NewHandler creates a new stream handler with default settings
func NewHandler() *Handler {
	return &Handler{
		HTTPClient: &http.Client{
			Timeout: 0, // No timeout for streaming
		},
		UserAgent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",

		ChunkSize:  10 * 1024 * 1024, // 10MB chunks
		MaxRetries: 3,
	}
}

// NewHandlerWithClient creates a handler with a custom HTTP client
func NewHandlerWithClient(client *http.Client) *Handler {
	h := NewHandler()
	h.HTTPClient = client
	return h
}

// Download downloads a format to a writer
func (h *Handler) Download(ctx context.Context, format types.Format, w io.Writer) error {
	if format.URL == "" {
		return fmt.Errorf("format has no URL")
	}

	// For formats with known content length, use range requests
	if format.ContentLength > 0 {
		return h.downloadWithRanges(ctx, format, w, 0, int64(format.ContentLength))
	}

	// Otherwise, simple download
	return h.downloadSimple(ctx, format.URL, w)
}

// DownloadRange downloads a specific byte range
func (h *Handler) DownloadRange(ctx context.Context, format types.Format, w io.Writer, start, end int64) error {
	if format.URL == "" {
		return fmt.Errorf("format has no URL")
	}

	return h.downloadWithRanges(ctx, format, w, start, end)
}

// GetStream returns a reader for streaming the format
func (h *Handler) GetStream(ctx context.Context, format types.Format) (io.ReadCloser, int64, error) {
	return h.GetStreamRange(ctx, format, 0, -1)
}

// GetStreamRange returns a reader for a specific byte range
func (h *Handler) GetStreamRange(ctx context.Context, format types.Format, start, end int64) (io.ReadCloser, int64, error) {
	if format.URL == "" {
		return nil, 0, fmt.Errorf("format has no URL")
	}

	req, err := http.NewRequestWithContext(ctx, "GET", format.URL, nil)
	if err != nil {
		return nil, 0, err
	}

	h.setHeaders(req)

	// Set range header
	if end > 0 {
		req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", start, end))
	} else if start > 0 {
		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", start))
	}

	resp, err := h.HTTPClient.Do(req)
	if err != nil {
		return nil, 0, err
	}

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
		resp.Body.Close()
		return nil, 0, fmt.Errorf("unexpected status: %d", resp.StatusCode)
	}

	return resp.Body, resp.ContentLength, nil
}

// downloadWithRanges downloads using chunked range requests
func (h *Handler) downloadWithRanges(ctx context.Context, format types.Format, w io.Writer, start, end int64) error {
	var downloaded int64 = start

	for downloaded < end {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		chunkEnd := downloaded + h.ChunkSize - 1
		if chunkEnd >= end {
			chunkEnd = end - 1
		}

		err := h.downloadChunk(ctx, format.URL, w, downloaded, chunkEnd)
		if err != nil {
			return err
		}

		downloaded = chunkEnd + 1
	}

	return nil
}

// downloadChunk downloads a single chunk with retries
func (h *Handler) downloadChunk(ctx context.Context, url string, w io.Writer, start, end int64) error {
	var lastErr error

	for attempt := 0; attempt < h.MaxRetries; attempt++ {
		if attempt > 0 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(time.Duration(attempt) * time.Second):
			}
		}

		err := h.doChunkRequest(ctx, url, w, start, end)
		if err == nil {
			return nil
		}

		lastErr = err
	}

	return fmt.Errorf("failed after %d retries: %w", h.MaxRetries, lastErr)
}

// doChunkRequest performs a single chunk request
func (h *Handler) doChunkRequest(ctx context.Context, url string, w io.Writer, start, end int64) error {
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return err
	}

	h.setHeaders(req)
	req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", start, end))

	resp, err := h.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusPartialContent && resp.StatusCode != http.StatusOK {
		return fmt.Errorf("unexpected status: %d", resp.StatusCode)
	}

	_, err = io.Copy(w, resp.Body)
	return err
}

// downloadSimple performs a simple download without range requests
func (h *Handler) downloadSimple(ctx context.Context, url string, w io.Writer) error {
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return err
	}

	h.setHeaders(req)

	resp, err := h.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("unexpected status: %d", resp.StatusCode)
	}

	_, err = io.Copy(w, resp.Body)
	return err
}

// setHeaders sets required headers for requests
func (h *Handler) setHeaders(req *http.Request) {
	req.Header.Set("User-Agent", h.UserAgent)
	req.Header.Set("Accept", "*/*")
	req.Header.Set("Accept-Language", "en-US,en;q=0.9")
	req.Header.Set("Origin", "https://www.youtube.com")
	req.Header.Set("Referer", "https://www.youtube.com/")
}

// StreamInfo contains metadata about a stream
type StreamInfo struct {
	ContentLength int64
	ContentType   string
	AcceptRanges  bool
}

// GetStreamInfo fetches stream metadata without downloading
func (h *Handler) GetStreamInfo(ctx context.Context, format types.Format) (*StreamInfo, error) {
	if format.URL == "" {
		return nil, fmt.Errorf("format has no URL")
	}

	req, err := http.NewRequestWithContext(ctx, "HEAD", format.URL, nil)
	if err != nil {
		return nil, err
	}

	h.setHeaders(req)

	resp, err := h.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status: %d", resp.StatusCode)
	}

	info := &StreamInfo{
		ContentLength: resp.ContentLength,
		ContentType:   resp.Header.Get("Content-Type"),
		AcceptRanges:  resp.Header.Get("Accept-Ranges") == "bytes",
	}

	// Try to get content length from header if not set
	if info.ContentLength <= 0 {
		if cl := resp.Header.Get("Content-Length"); cl != "" {
			info.ContentLength, _ = strconv.ParseInt(cl, 10, 64)
		}
	}

	return info, nil
}

// Progress tracks download progress
type Progress struct {
	Total      int64
	Downloaded int64
	Speed      float64 // bytes per second
}

// ProgressCallback is called with download progress updates
type ProgressCallback func(Progress)

// DownloadWithProgress downloads with progress reporting
func (h *Handler) DownloadWithProgress(ctx context.Context, format types.Format, w io.Writer, callback ProgressCallback) error {
	if format.URL == "" {
		return fmt.Errorf("format has no URL")
	}

	total := int64(format.ContentLength)
	if total <= 0 {
		info, err := h.GetStreamInfo(ctx, format)
		if err == nil && info.ContentLength > 0 {
			total = info.ContentLength
		}
	}

	pw := &progressWriter{
		writer:    w,
		total:     total,
		callback:  callback,
		startTime: time.Now(),
	}

	if total > 0 {
		return h.downloadWithRanges(ctx, format, pw, 0, total)
	}

	return h.downloadSimple(ctx, format.URL, pw)
}

// progressWriter wraps a writer to track progress
type progressWriter struct {
	writer     io.Writer
	total      int64
	downloaded int64
	callback   ProgressCallback
	startTime  time.Time
}

func (pw *progressWriter) Write(p []byte) (int, error) {
	n, err := pw.writer.Write(p)
	if err != nil {
		return n, err
	}

	pw.downloaded += int64(n)

	if pw.callback != nil {
		elapsed := time.Since(pw.startTime).Seconds()
		speed := float64(pw.downloaded) / elapsed

		pw.callback(Progress{
			Total:      pw.total,
			Downloaded: pw.downloaded,
			Speed:      speed,
		})
	}

	return n, nil
}
